Zend/zend_execute.c

```diff
@@ -177,6 +177,29 @@ ZEND_API void* zend_vm_stack_extend(size_t size)
 	return ptr;
 }
 
+zend_bool zend_vm_is_fetching_reference(const zend_op * opline) {
+	zend_op const *next = opline + 1;
+
+	while (next->opcode == ZEND_NOP ||
+			next->opcode == ZEND_TICKS ||
+			next->opcode == ZEND_EXT_NOP ||
+			next->opcode == ZEND_EXT_STMT)
+		next++;
+
+	switch (next->opcode) {
+		case ZEND_SEND_REF:
+		case ZEND_RETURN_BY_REF:
+		case ZEND_ASSIGN_REF:
+		case ZEND_INIT_ARRAY:	
+		case ZEND_ADD_ARRAY_ELEMENT: 
+		case ZEND_YIELD: {
+			return 1;
```

Comment:
I'll look at these, thanks ...

Cheers
Joe

On Wed, Apr 6, 2016 at 10:33 AM, Dmitry Stogov <notifications@github.com>
wrote:

> In Zend/zend_execute.c
> <https://github.com/php/php-src/pull/1797#discussion_r58676887>:
>
> > +	zend_op const *next = opline + 1;
> > +
> > +	while (next->opcode == ZEND_NOP ||
> > +			next->opcode == ZEND_TICKS ||
> > +			next->opcode == ZEND_EXT_NOP ||
> > +			next->opcode == ZEND_EXT_STMT)
> > +		next++;
> > +
> > +	switch (next->opcode) {
> > +		case ZEND_SEND_REF:
> > +		case ZEND_RETURN_BY_REF:
> > +		case ZEND_ASSIGN_REF:
> > +		case ZEND_INIT_ARRAY:	
> > +		case ZEND_ADD_ARRAY_ELEMENT:
> > +		case ZEND_YIELD: {
> > +			return 1;
>
> These checks are not enough:
>
> <?php
> class A {
> public int $x = 5;
> }
> $a = new A;
> $b = [1];
> $a->x =& $b[0];
> $b[0] = "opd";
> var_dump($a);'
> ?>
>
> object(A)#1 <https://github.com/php/php-src/pull/1> (1) {
> ["x"]=>
> &string(3) "opd"
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/php/php-src/pull/1797/files/46742861d254a5fe74b6d88dde4232e4bae4bb5c#r58676887>
>